### PR TITLE
Add @mcap/support, @mcap/nodejs, and @mcap/browser libraries

### LINKED ISSUE #719: Publish decompression handler boilerplate to NPM
Some boilerplate is required to set up decompression handlers when reading an MCAP file. In https://github.com/foxglove/ws-protocol/pull/270 I copy/pasted some of the code we use for this from another repo. We have similar code both here and in foxglove/studio. These should be published to NPM so they can be reused.

Library name ideas:
- `@mcap/compression-support`
- `@mcap/wasm-compression-support`

### LINKED ISSUE #720: Publish Node fs IReadable/IWritable boilerplate to NPM
Some boilerplate is required to make an IReadable/IWritable from a fs handle when reading an MCAP file in Node.js. In https://github.com/foxglove/ws-protocol/pull/270 I copy/pasted some of the code we use for this from another script. This IReadable implementation should be published to NPM so it can be reused.

Library name ideas:

- `@mcap/node`
- `@mcap/node-support`
- `@mcap/node-fs-support`

### PR DESCRIPTION
**Pull request by jtbandes**
**Public-Facing Changes**
- [TypeScript] Added a new package `@mcap/support`. This package provides decompression functions for zstd, lz4, and bz2, and utility functions for working with protobuf descriptors.
- [TypeScript] Added a new package `@mcap/nodejs`. This package provides IReadable and IWritable implementations for Node.js `FileHandle`, reducing boilerplate for reading or writing MCAP files.
- [TypeScript] Added a new package `@mcap/browser`. This package provides an IReadable implementation for the browser `Blob` (`File`) type, reducing boilerplate for reading MCAP files.
- [TypeScript] Examples have been updated to use `@mcap/support` + `@mcap/nodejs` and remove repeated boilerplate

**Description**

**Note:** This began as un-revert of #838 (see #841)
Relates to FG-2274
Resolves #719 
Resolves #720 

This package has been migrated from the @foxglove/mcap-support internal package in the foxglove/studio repo. Two new additions are `FileHandleWritable` under the `@mcap/nodejs` package, and `protobufDescriptors.ts` to provide helper methods for protobuf descriptors and avoid each project having to include a `protobufjs.d.ts` copy-paste.

~Studio's protobuf timestamp/duration rewriting logic has been replaced with `processRootType` & `processMessageDefinitions` functions that Studio can continue using to implement that behavior without inflicting its opinion upon other users of the library.~

Porting of deserialization for different encodings will be done in a separate PR when we have a chance to revisit the reader APIs to make them more friendly.

### CODE REVIEW

**Comment by defunctzombie** [comment]
I'd like to see a readme file with some documentation on how to use this stuff in the various settings. If we are going to take the time to split this out and have this library docs should be a part of that workflow.

**Comment by jtbandes** [comment]
New docs preview build: https://f816178c.mcap.pages.dev/docs/typescript/

File: typescript/browser/LICENSE [hunk 1]
```diff
@@ -0,0 +1,21 @@
+MIT License
```
Comment (snosenzo) [inline typescript/browser/LICENSE:1]: fyi at least for the omgidl repo I was told that it's fine to just have a top level repo license and it doesn't need to be included in all subsequent workspaces

**Review by snosenzo — APPROVED** [review]
I think it looks good to me. I don't have any issues with the API, the examples are clear. Only thing you might want to change is to not include test files in the build. To do this usually I have to make a `tsconfig.build.js` that excludes test file endings

File: typescript/CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+```
+
+## Releasing to NPM
+
+- Check out the version of the code you want to release
+- Update package.json in `typescript/{pkg}/package.json` with the new version.
+- Make a PR with your changes to package.json
+- Wait for the PR to pass CI and merge
+- Checkout main and tag the merged commit with `releases/typescript/{pkg}/v#.#.#` (replace #.#.# with the version you used in package.json)
+- Push the new tag to the repo with `git push origin releases/typescript/{pkg}/v#.#.#`
```
Comment (snosenzo) [inline typescript/CONTRIBUTING.md:42]: don't tag references have to be preceded by `refs/tags/`?

Comment (jtbandes) [reply typescript/CONTRIBUTING.md:42]: I think you can for disambiguation, but it's not required

**Comment by jhurliman** [comment]
Any progress on this? Making a small command-line utility for reading/writing MCAP files is currently more verbose in TypeScript than its C++ equivalent.

**Comment by defunctzombie** [comment]
> Any progress on this? Making a small command-line utility for reading/writing MCAP files is currently more verbose in TypeScript than its C++ equivalent.

It's overall in a good place but on my list of stuff to review. We have some other work that's higher priority so this hasn't gotten eyeballs yet.

**Comment by jtbandes** [comment]
It'll still be fairly verbose with this package in its current form, but still better than it was. There are more improvements we could make that would let you basically "readMessages()" to get fully deserialized messages, but haven't decided yet whether that is going to block this merging or not.

**Review by defunctzombie — APPROVED** [review]
I like the `@mcap/node` and `@mcap/browser` packages along with the decompression helpers those offer. I think that `parseChannel` is too high level and specific to Studio to bundle as a one-size solution and am less keep on it depending on all the various serialization formats. Also less excited about the external maintenance that brings with it.

My opinion is that `@mcap/core` readers should offer a better interface for managing message deserialization and that using `parseChannel` is more unpleasant.

My take on this PR is that we should land the node and browser packages which already help usability on their own, and then look at how reading interfaces can aid with schema and message deserialization that has even less boilerplate than the parseChannel approach. Ultimately I will leave it for you to decide on which parts of this PR to merge but I would not offer parseChannel as something we should support in its current form.

**Comment by jtbandes** [comment]
@defunctzombie the parseChannel stuff has been stripped out – is this in a place you are happy with now?

**Comment by jtbandes** [comment]
":shipit:" – said offline

**Pull request merged**